File: chapter-04/quicksort/quicksort.py
def quicksort(arr: list[int]) -> list[int]:
    if len(arr) < 2:
        return arr

    left_array = []
    right_array = []

    pivot = arr[0]
    for num in arr[1:]:
        if pivot >= num:
            left_array.append(num)
        elif pivot < num:
            right_array.append(num)
    
    return quicksort(left_array) + [pivot] + quicksort(right_array)


lista = [9, 8, 7, 6, 5, 4, 3, 2, 1, 0]
print(quicksort(lista))

File: chapter-04/quicksort/quicksort.go
package main

import (
	"fmt"
	"time"
)

func Quicksort(arr []int) []int {
	if len(arr) < 2 {
		return arr
	}

	lesser := []int{}
	greater := []int{}

	pivot := arr[0]

	for _, num := range arr[1:] {
		if pivot >= num {
			lesser = append(lesser, num)
		} else if pivot < num {
			greater = append(greater, num)
		}
	}

	return append(append(Quicksort(lesser), pivot), Quicksort(greater)...)
}

func main() {
	array := []int{}
	for x := 1000; x > 0; x-- {
		array = append(array, x)
	}

	startTime := time.Now()
	sortedArray := Quicksort(array)
	fmt.Printf("The sorted array is: %v\nIt took the program %v to sort it.\n", sortedArray, time.Since(startTime))
}
